packages/dialog/test/header-footer.test.js

```diff
@@ -0,0 +1,429 @@
+import { fixtureSync } from '@vaadin/testing-helpers';
+import '../vaadin-dialog.js';
+
+describe('header/footer feature', () => {
+  let dialog, overlay;
+
+  beforeEach(() => {
+    dialog = fixtureSync('<vaadin-dialog></vaadin-dialog>');
+    overlay = dialog.$.overlay;
+  });
+
+  afterEach(() => {
+    dialog.opened = false;
+  });
+
+  describe('vaadin-dialog header-title attribute', () => {
+    const HEADER_TITLE = '__HEADER_TITLE__';
+
+    it('should not have title element if header-title is not set', () => {
+      dialog.opened = true;
+
+      expect(overlay.querySelector('#title')).to.not.exist;
+    });
+
+    it('should render header-title when set', () => {
+      dialog.headerTitle = HEADER_TITLE;
+      dialog.opened = true;
+
+      expect(overlay.querySelector('#title')).to.exist;
+      expect(overlay.textContent).to.include(HEADER_TITLE);
+    });
+
+    it('should remove title element if header-title is unset', () => {
+      dialog.headerTitle = HEADER_TITLE;
+      dialog.opened = true;
+
+      dialog.headerTitle = null;
+      expect(overlay.querySelector('#title')).to.not.exist;
+    });
+
+    it('should remove title element if header-title is set to empty string', () => {
+      dialog.headerTitle = HEADER_TITLE;
+      dialog.opened = true;
+
+      dialog.headerTitle = '';
+      expect(overlay.querySelector('#title')).to.not.exist;
+    });
+
+    it('should not have [has-title] attribute on overlay element if header-title is not set', () => {
+      dialog.opened = true;
+
+      expect(overlay.hasAttribute('has-title')).to.be.not.ok;
+    });
+
+    it('should add [has-title] attribute on overlay element if header-title is set', () => {
+      dialog.headerTitle = HEADER_TITLE;
+      dialog.opened = true;
+
+      expect(overlay.hasAttribute('has-title')).to.be.ok;
+    });
+
+    it('should remove [has-title] attribute on overlay element if header-title is set', () => {
+      dialog.headerTitle = HEADER_TITLE;
+      dialog.opened = true;
+
+      dialog.headerTitle = null;
+
+      expect(overlay.hasAttribute('has-title')).to.be.not.ok;
+    });
+
+    it('[part=header] should have display:none if no header-title is set', () => {
+      dialog.opened = true;
+
+      expect(getComputedStyle(overlay.shadowRoot.querySelector('[part=header]')).display).to.be.equal('none');
+    });
+
+    it('[part=header] should have be displayed if no header-title is set', () => {
+      dialog.headerTitle = HEADER_TITLE;
+      dialog.opened = true;
+
+      expect(getComputedStyle(overlay.shadowRoot.querySelector('[part=header]')).display).to.not.be.equal('none');
+    });
+  });
+
+  describe('vaadin-dialog headerRenderer', () => {
+    const HEADER_CONTENT = '__HEADER_CONTENT__';
+    const headerRenderer = (root) => {
+      if (root.firstChild) {
+        return;
+      }
+
+      const span = document.createElement('span');
+      span.textContent = HEADER_CONTENT;
+      root.appendChild(span);
+    };
```

Comment:
There are 11 occurrences of this renderer function in the test suite, probably worth extracting to helper:

```js
function createRenderer(text) {
  return (root) => {
      if (root.firstChild) {
        return;
      }

      const span = document.createElement('span');
      span.textContent = text;
      root.appendChild(span);
    };
}

const headerRenderer = createRenderer(HEADER_CONTENT);
```